draft-ietf-hpke-pq.md

```diff
@@ -148,36 +148,45 @@ The NIST Module-Lattice-Based Key-Encapsulation Mechanism is defined in
 {{FIPS203}}.  In this section, we define how to implement the HPKE KEM interface
 using ML-KEM.
 
-The HPKE `DeriveKeyPair` function corresponds to the function
-`ML-KEM.KeyGen_internal` in {{FIPS203}}.  The input `ikm` MUST be exactly
-`Nsk = 64` bytes long.  The `d` and `z` inputs to `ML-KEM.KeyGen_internal` are
-the first and last 32-byte segments of `ikm`, respectively.  The output `skX` is
-the generated decapsulation key and the output `pkX` is the generated
-encapsulation key.
+The HPKE `DeriveKeyPair` function uses the SHAKE256 KDF (see
+{{single-stage-kdfs}}) to derive an ML-KEM decapsulation key in the 64-byte seed
+format, then uses the function `ML-KEM.KeyGen_internal` from {{FIPS203}} to
+compute the corresponding encapsulation key.
 
 ~~~ pseudocode
+def expandDecapsKey(dk):
+    d = dk[:32]
+    z = dk[32:]
+    (ek, expanded_dk) = ML-KEM.KeyGen_internal(d, z)
+    return (expanded_dk, ek)
+
 def DeriveKeyPair(ikm):
-    if len(ikm) != 64:
+    if len(ikm) < 32:
         raise DeriveKeyPairError
 
-    d = ikm[:32]
-    z = ikm[32:]
-
-    dk = ikm
-    (ek, _) = ML-KEM.KeyGen_internal(d, z)
+    dk = SHAKE256.LabeledDerive(ikm, "ML-KEM DeriveKeyPair", "", 64)
+    (_expanded_dk, ek) = expandDecapsKey(dk)
     return (dk, ek)
 ~~~
 
-The `GenerateKeyPair` function is simply `DeriveKeyPair` with a pseudorandom
-`ikm` value.  As long as the bytes supplied by `random` meet the randomness
-requirements of {{FIPS203}}, this corresponds to the `ML-KEM.KeyGen` function,
-with the distinction that the decapsulation key is returned in seed format
-rather than the expanded form returned by `ML-KEM.KeyGen`.
+As discussed in {{HPKE}}, the value of `suite_id` used within
+LabeledDerive identifies the KEM in use:
+
+* ML-KEM-512: `KEM\x00\x40` (hex: 4b454d0040)
+* ML-KEM-768: `KEM\x00\x41` (hex: 4b454d0041)
+* ML-KEM-1024: `KEM\x00\x42` (hex: 4b454d0042)
+
+The `GenerateKeyPair` function simply calls `ML-KEM.KeyGen_internal` with a
+pseudorandom `dk` value.  As long as the bytes supplied by `random` meet the
+randomness requirements of {{FIPS203}}, this corresponds to the `ML-KEM.KeyGen`
+function, with the distinction that the decapsulation key is returned in seed
+format rather than the expanded form returned by `ML-KEM.KeyGen`.
 
 ~~~ pseudocode
 def GenerateKeyPair():
-    dz = random(64)
-    return DeriveKeyPair(dz)
+    dk = random(64)
+    (_expanded_dk, ek) = expandDecapsKey(dk)
+    return (dk, ek)
 ~~~
 
 The `SerializePublicKey` and `DeserializePublicKey` functions are both the
```

Comment:
I think you are missing `SerializePrivateKey` and `DeserializePrivateKey` descriptions, similarly trivial.